ext/standard/array.c

```diff
@@ -345,7 +345,24 @@ static zend_always_inline int php_array_key_compare_string_locale_unstable_i(Buc
 
 static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucket *s) /* {{{ */
 {
-	return zend_compare(&f->val, &s->val);
+	int result = zend_compare(&f->val, &s->val);
+	// Special handling for enums
+	if (result == ZEND_UNCOMPARABLE) {
+		zval *rhs = &s->val;
+		if (UNEXPECTED(Z_TYPE_P(rhs) == IS_OBJECT && (Z_OBJCE_P(rhs)->ce_flags & ZEND_ACC_ENUM))) {
+			zval *lhs = &f->val;
+			if (Z_TYPE_P(lhs) == IS_OBJECT && (Z_OBJCE_P(lhs)->ce_flags & ZEND_ACC_ENUM)) {
+				// Order doesn't matter, we just need to group the same enum values
+				uintptr_t lhs_uintptr = (uintptr_t)Z_OBJ_P(lhs);
+				uintptr_t rhs_uintptr = (uintptr_t)Z_OBJ_P(rhs);
+				return lhs_uintptr == rhs_uintptr ? 0 : (lhs_uintptr < rhs_uintptr ? -1 : 1);
+			} else {
+				// Shift enums to the end of the array
```

Comment:
what is the reason for this else?

Reply:
It's just personal preference. I prefer keeping the nesting when the `if` isn't an intentional error-handling early-return.

Reply:
I mean if we remove it, the result will be `ZEND_UNCOMPARABLE` - is there any problem with it? The effect should be the same, but no forced order.

Reply:
maybe return `-ZEND_UNCOMPARABLE` to stress that?

Reply:
I'm not sure if "not comparable, negated" makes more sense than `-1` which means keep the order of the operands. The value of `ZEND_UNCOMPARABLE` should be changed anyway to allow for an actual distinction between "flip the operands" and "this comparison doesn't make sense".

Reply:
I am looking at the code and a few lines there is a comment the order does not matter.

If it does not matter, then the `else` can be completely removed and fallback to `ZEND_UNCOMPARABLE`

If it does matter, then returning a compare result of ptr is problematic, as ptr is random, thus the compare result will be random on individual runs.

Reply:
I agree the comment is confusing. What it means to say is that the actual order of the enum values doesn't matter (i.e. which ones come first) but it does matter that the order is consistent. I'll improve the comment before merging.